### PR TITLE
[FLOC-4245] Factor out plugin loading.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-4245

### CODE REVIEW

**Comment by sarum90** [comment]
Potentially add a bit more context on either the PR comment or the JIRA ticket. I happen to know the motivation, but it'd be nice to be able to know that just by reading the PR description.

File: flocker/node/script.py [hunk 1]
```diff
@@ -552,8 +404,6 @@ def get_api(backend, api_args, reactor, cluster_id):
 
 class AgentService(PClass):
     """
-    :ivar backends: ``BackendDescription`` instances describing how to use each
```
Comment (sarum90) [inline flocker/node/script.py]: I think this still needs documentation, just different documentation.

File: flocker/node/script.py [hunk 2]
```diff
@@ -365,123 +358,6 @@ def get_configuration(options):
...
-        ZFS where nodes interact directly with each other for data movement.
-    :ivar block: The Infrastructure-as-a-Service deployer - suitable for use
-        with system like EBS where volumes can be attached to nodes as block
-        devices and then detached (and then re-attached to other nodes).
-    """
-    p2p = NamedConstant()
-    block = NamedConstant()
-
-
-class BackendDescription(PClass):
```
Comment (sarum90) [inline flocker/node/script.py]: :question: No action required. Just documenting my review process:

Confirmed that no one was incorrectly import BackendDescription from this location (https://github.com/search?utf8=%E2%9C%93&q=%22flocker.node.script%22+extension%3Apy&type=Code&ref=searchresults) So it should be safe to move.

File: flocker/common/plugin.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    :attr str plugin_name: Name of the plugin looked for.
+    """
+    def __str__(self):
+        return (
+            "'{!s}' is neither a built-in plugin nor a 3rd party "
+            "module.".format(self.plugin_name)
+        )
+
+
+class InvalidPlugin(Exception, object):
```
Comment (sarum90) [inline flocker/common/plugin.py]: :question: What's the purpose of also subclassing object here?

File: flocker/common/plugin.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                        actual_type=type(builtin),
+                    )
+                )
+
+        return (True, "")
+
+    def get(self, plugin_name):
+        """
+        Find the backend in ``backends`` that matches the one named by
+        ``backend_name``. If not found then attempt is made to load it as
```
Comment (sarum90) [inline flocker/common/plugin.py]: s/then attempt/then an attempt/

**Comment by sarum90** [comment]
Couple nits, but in general looks great.

Nice to get this code refactored to a location that it can be re-used by other parts of the flocker codebase.

File: flocker/common/plugin.py (hunk 3 above)
Comment (tomprince) [inline flocker/common/plugin.py]: I added that when trying to debug https://github.com/hynek/characteristic/issues/37. It isn't needed.

File: flocker/node/script.py (hunk 1 above)
Comment (tomprince) [inline flocker/node/script.py]: The original refactoring didn't have an attribute here. I forgot to add back the documentation when I added it back.

**Pull request merged**